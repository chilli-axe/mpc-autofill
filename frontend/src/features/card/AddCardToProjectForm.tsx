import React, { FormEvent, useState } from "react";
import Button from "react-bootstrap/Button";
import Col from "react-bootstrap/Col";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";

import {
  ProjectMaxSize,
  ReversedCardTypePrefixes,
  SelectedImageSeparator,
} from "@/common/constants";
import {
  convertLinesIntoSlotProjectMembers,
  processLine,
} from "@/common/processing";
import { CardDocument, useAppDispatch, useAppSelector } from "@/common/types";
import { RightPaddedIcon } from "@/components/icon";
import { useGetDFCPairsQuery } from "@/store/api";
import { addMembers, selectProjectSize } from "@/store/slices/projectSlice";
import { selectFuzzySearch } from "@/store/slices/searchSettingsSlice";
import { setNotification } from "@/store/slices/toastsSlice";

export function AddCardToProjectForm({
  cardDocument,
}: {
  cardDocument: CardDocument;
}) {
  const dispatch = useAppDispatch();
  const dfcPairsQuery = useGetDFCPairsQuery();
  const fuzzySearch = useAppSelector(selectFuzzySearch);

  const projectSize = useAppSelector(selectProjectSize);
  const maxQuantity = ProjectMaxSize - projectSize;
  const atMaxCapacity = maxQuantity === 0;
  const [quantity, setQuantity] = useState<number>(1);
  const handleAddToProject = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault(); // to prevent reloading the page

    const prefix = ReversedCardTypePrefixes[cardDocument.cardType];
    const lineString = `${quantity} ${prefix}${cardDocument.searchq}${SelectedImageSeparator}${cardDocument.identifier}`;
    const line = processLine(lineString, dfcPairsQuery.data ?? {}, fuzzySearch);

    dispatch(
      addMembers({
        members: convertLinesIntoSlotProjectMembers([line], projectSize),
      })
    );
    dispatch(
      setNotification([
        Math.random().toString(),
        {
          name: "Added to Project",
          message: `Added ${quantity} ${quantity > 1 ? "copies" : "copy"} of ${
            cardDocument.name
          } to your project!`,
          level: "info",
        },
      ])
    );
    setQuantity(1);
  };

  return (
    <Form onSubmit={handleAddToProject}>
      <Row className="g-0 pt-3">
        <Col lg={2} md={3}>
          <Form.Control
            type="number"
            pattern="[0-9]*"
            placeholder="1"
            min={1}
            max={maxQuantity}
            value={atMaxCapacity ? 0 : quantity}
            disabled={atMaxCapacity}
            alt="Quantity of card to add to project"
            onChange={(event) => {
              if (event.target.value) {
                setQuantity(parseInt(event.target.value));
              }
            }}
          />
        </Col>
        <Col lg={10} md={9}>
          <div className="d-grid gap-0">
            <Button variant="success" type="submit" disabled={atMaxCapacity}>
              <RightPaddedIcon bootstrapIconName="plus-circle" /> Add to Project
            </Button>
          </div>
        </Col>
      </Row>
    </Form>
  );
}
