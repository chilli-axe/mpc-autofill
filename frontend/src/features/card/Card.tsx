/**
 * This component is a fundamental UI building block for displaying cards.
 * Displays a card's image, some extra information (its name, its source's name, and its DPI),
 * and has optional props for extending the component to include extra functionality.
 * If being used in a gallery, the previous and next images can be cached for visual smoothness.
 */

import { OnLoadingComplete } from "next/dist/shared/lib/get-img-props";
import Image from "next/image";
import React, {
  memo,
  PropsWithChildren,
  ReactElement,
  useEffect,
  useRef,
  useState,
} from "react";
import BSCard from "react-bootstrap/Card";
import Col from "react-bootstrap/Col";
import styled from "styled-components";

import { SearchQuery, useAppDispatch, useAppSelector } from "@/common/types";
import { CardDocument } from "@/common/types";
import { Spinner } from "@/components/Spinner";
import { selectCardDocumentByIdentifier } from "@/store/slices/cardDocumentsSlice";
import { setSelectedCardAndShowModal } from "@/store/slices/modalsSlice";
import { RootState } from "@/store/store";

const HiddenImage = styled(Image)`
  z-index: 0;
  opacity: 0;
`;

const VisibleImage = styled(Image)<{
  imageIsLoading?: boolean;
  showDetailedViewOnClick?: boolean;
  zIndex?: number;
}>`
  z-index: ${(props) => props.zIndex ?? 1};
  &:hover {
    cursor: ${(props) => (props.showDetailedViewOnClick ? "pointer" : "auto")};
  }
  opacity: ${(props) => (props.imageIsLoading ? 0 : 1)};
`;

const OutlinedBSCardSubtitle = styled(BSCard.Subtitle)`
  outline: solid 1px #ffffff00;
  transition: outline 0.2s ease-in-out;
  &:hover {
    outline-color: #ffffffff;
    cursor: pointer;
  }
`;

export function getImageKey(
  cardDocument: CardDocument,
  small: boolean
): string {
  return `${cardDocument.identifier}-${
    small ? "small" : "large"
  }-${cardDocument.sourceType?.toLowerCase().replace(" ", "_")}`;
}

interface CardImageProps {
  maybeCardDocument: CardDocument | null;
  hidden: boolean;
  small: boolean;
  showDetailedViewOnClick: boolean;
}

function CardImage({
  maybeCardDocument,
  hidden,
  small,
  showDetailedViewOnClick,
}: CardImageProps) {
  //# region queries and hooks

  const dispatch = useAppDispatch();

  //# endregion

  //# region state

  const [imageState, setImageState] = useState<
    | "loading-from-bucket"
    | "loading-from-fallback"
    | "loaded-from-bucket"
    | "loaded-from-fallback"
    | "errored"
  >("loading-from-bucket");
  const image = useRef<HTMLImageElement>(null);

  //# endregion

  //# region callbacks

  const onLoadingComplete: OnLoadingComplete = (img) => {
    if (imageState === "loading-from-bucket") {
      setImageState("loaded-from-bucket");
    } else if (imageState === "loading-from-fallback") {
      setImageState("loaded-from-fallback");
    }
  };
  const onError: React.ReactEventHandler<HTMLImageElement> = (img) => {
    img.preventDefault();
    img.currentTarget.onerror = null;
    setImageState((value) =>
      value === "loading-from-bucket" || value === "loaded-from-bucket"
        ? "loading-from-fallback"
        : "errored"
    );
  };
  const handleShowDetailedView = () => {
    if (showDetailedViewOnClick && maybeCardDocument != null) {
      dispatch(
        setSelectedCardAndShowModal([maybeCardDocument, "cardDetailedView"])
      );
    }
  };

  //# endregion

  //# region effects

  /**
   * Ensure that the small thumbnail fades in each time the selected image changes.
   * Next.js seems to not fire `onLoadingComplete` when opening a page with a cached image.
   * This implementation was retrieved from https://stackoverflow.com/a/59809184
   */
  useEffect(() => {
    setImageState(
      image.current == null || !image.current.complete
        ? "loading-from-bucket"
        : "loaded-from-bucket"
    );
  }, [maybeCardDocument?.identifier]);

  //# endregion

  //# region computed constants

  // attempt to load directly from bucket first
  const imageBucketURL = process.env.NEXT_PUBLIC_IMAGE_BUCKET_URL;
  const imageBucketURLValid =
    imageBucketURL != null && !!maybeCardDocument?.sourceType;

  const loadFromBucket =
    imageBucketURLValid &&
    (imageState === "loading-from-bucket" ||
      imageState === "loaded-from-bucket");
  const imageKey = maybeCardDocument && getImageKey(maybeCardDocument, small);
  const thumbnailBucketURL = `${imageBucketURL}/${imageKey}`;

  // if image is unavailable in bucket, fall back on loading from worker if possible
  const imageWorkerURL = process.env.NEXT_PUBLIC_IMAGE_WORKER_URL;
  const imageWorkerURLValid =
    imageWorkerURL != null && !!maybeCardDocument?.sourceType;

  const smallThumbnailURL = imageWorkerURLValid
    ? `${imageWorkerURL}/images/google_drive/small/${maybeCardDocument?.identifier}.jpg`
    : maybeCardDocument?.smallThumbnailUrl;
  const mediumThumbnailURL = imageWorkerURLValid
    ? `${imageWorkerURL}/images/google_drive/large/${maybeCardDocument?.identifier}.jpg`
    : maybeCardDocument?.mediumThumbnailUrl;
  const thumbnailFallbackURL = small ? smallThumbnailURL : mediumThumbnailURL;
  const imageSrc = loadFromBucket ? thumbnailBucketURL : thumbnailFallbackURL;

  // if loading from fallback fails, display a 404 error image
  const errorImageSrc = small ? "/error_404.png" : "/error_404_med.png";

  // a few other computed constants
  const imageAlt = maybeCardDocument?.name ?? "Unnamed Card";
  const imageIsLoading =
    imageState === "loading-from-bucket" ||
    imageState === "loading-from-fallback";
  const showSpinner = imageIsLoading && !hidden;

  //# endregion

  return (
    <>
      {showSpinner && <Spinner zIndex={2} />}
      {imageSrc != null &&
        (hidden ? (
          <HiddenImage
            ref={image}
            className="card-img"
            loading="lazy"
            src={imageSrc}
            onLoadingComplete={onLoadingComplete}
            onErrorCapture={onError}
            alt={imageAlt}
            fill={true}
          />
        ) : (
          <>
            {imageState === "errored" ? (
              <VisibleImage
                ref={image}
                className="card-img card-img-fade-in"
                loading="lazy"
                src={errorImageSrc}
                alt={""}
                fill={true}
              />
            ) : (
              <VisibleImage
                ref={image}
                className="card-img card-img-fade-in"
                loading="lazy"
                imageIsLoading={imageIsLoading}
                showDetailedViewOnClick={showDetailedViewOnClick}
                src={imageSrc}
                onLoadingComplete={onLoadingComplete}
                onErrorCapture={onError}
                onClick={handleShowDetailedView}
                alt={imageAlt}
                fill={true}
              />
            )}
          </>
        ))}
    </>
  );
}

export const MemoizedCardImage = memo(CardImage);

interface CardProportionWrapperProps {
  small: boolean;
  bordered?: boolean;
}

const CardProportionWrapperStyle = styled.div<{ $borderWidth?: number }>`
  z-index: 0;
  background: #4e5d6c;
  border: solid ${(props) => props.$borderWidth ?? 0}px black;
`;

function CardProportionWrapper({
  small,
  bordered = false,
  children,
}: PropsWithChildren<CardProportionWrapperProps>) {
  return (
    <CardProportionWrapperStyle
      $borderWidth={bordered ? 2 : 0}
      className={`rounded-${small ? "lg" : "xl"} shadow-lg ratio ratio-7x5`}
    >
      {children}
    </CardProportionWrapperStyle>
  );
}

export const MemoizedCardProportionWrapper = memo(CardProportionWrapper);

interface CardProps {
  /** The card image identifier to display. */
  maybeCardDocument: CardDocument | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the previous image for visual smoothness. */
  maybePreviousCardDocument?: CardDocument | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the next image for visual smoothness. */
  maybeNextCardDocument?: CardDocument | undefined;
  /** The string to display in the `Card` header. */
  cardHeaderTitle: string;
  /** An element (intended for use with a series of buttons) to include in the `Card` header.  */
  cardHeaderButtons?: ReactElement;
  /** An element (e.g. prev/next buttons) to display in the card footer. If not passed, no footer will be rendered. */
  cardFooter?: ReactElement;
  /** A callback function for when the `Card` (the HTML surrounding the image) is clicked. */
  cardOnClick?: React.MouseEventHandler<HTMLElement>;
  /** A callback function for when the card name is clicked. */
  nameOnClick?: React.MouseEventHandler<HTMLElement>;
  /** The `SearchQuery` specified when searching for this card. */
  searchQuery?: SearchQuery | undefined;
  /** Whether no search results were found when searching for `searchQuery` under the configured search settings. */
  noResultsFound: boolean;
  /** Whether to highlight this card by showing a glowing border around it. */
  highlight?: boolean;
}

/**
 * This component enables displaying cards with auxiliary information in a flexible, consistent way.
 */
export function Card({
  maybeCardDocument,
  maybePreviousCardDocument,
  maybeNextCardDocument,
  cardHeaderTitle,
  cardHeaderButtons,
  cardFooter,
  cardOnClick,
  nameOnClick,
  searchQuery,
  noResultsFound,
  highlight,
}: CardProps) {
  //# region computed constants

  const cardImageElements =
    maybeCardDocument != null ? (
      <>
        <MemoizedCardImage
          maybeCardDocument={maybeCardDocument}
          hidden={false}
          small={true}
          showDetailedViewOnClick={cardOnClick == null}
        />
        {maybePreviousCardDocument != null &&
          maybePreviousCardDocument.identifier !==
            maybeCardDocument?.identifier && (
            <MemoizedCardImage
              maybeCardDocument={maybePreviousCardDocument}
              hidden={true}
              small={true}
              showDetailedViewOnClick={false}
            />
          )}
        {maybeNextCardDocument != null &&
          maybeNextCardDocument.identifier !==
            maybeCardDocument?.identifier && (
            <MemoizedCardImage
              maybeCardDocument={maybeNextCardDocument}
              hidden={true}
              small={true}
              showDetailedViewOnClick={false}
            />
          )}
      </>
    ) : noResultsFound ? (
      <Image
        className="card-img card-img-fade-in"
        loading="lazy"
        style={{ zIndex: 1 }}
        src="/blank.png"
        alt="Card not found"
        fill={true}
      />
    ) : (
      <Spinner />
    );
  const BSCardSubtitle: typeof BSCard.Subtitle =
    nameOnClick != null ? OutlinedBSCardSubtitle : BSCard.Subtitle;

  //# endregion

  return (
    <BSCard
      className={`mpccard ${highlight ? "mpccard-highlight" : "mpccard-hover"}`}
      onClick={cardOnClick}
      style={{ contentVisibility: "auto" }}
    >
      <BSCard.Header className="pb-0 text-center">
        <p className="mpccard-slot">{cardHeaderTitle}</p>
        {cardHeaderButtons}
      </BSCard.Header>
      <div>
        <MemoizedCardProportionWrapper small={true}>
          {cardImageElements}
        </MemoizedCardProportionWrapper>
        <BSCard.Body className="mb-0 text-center">
          <BSCardSubtitle className="mpccard-name" onClick={nameOnClick}>
            {maybeCardDocument != null && maybeCardDocument.name}
            {maybeCardDocument == null &&
              searchQuery != undefined &&
              searchQuery.query}
          </BSCardSubtitle>
          <div className="mpccard-spacing">
            <BSCard.Text className="mpccard-source">
              {maybeCardDocument != null &&
                `${maybeCardDocument.sourceVerbose} [${maybeCardDocument.dpi} DPI]`}
              {maybeCardDocument == null &&
                searchQuery != undefined &&
                "Your search query"}
            </BSCard.Text>
          </div>
        </BSCard.Body>
      </div>
      {cardFooter != null && (
        <BSCard.Footer
          className="padding-top"
          style={{ paddingTop: 50 + "px" }}
        >
          {cardFooter}
        </BSCard.Footer>
      )}
    </BSCard>
  );
}

export const MemoizedCard = memo(Card);

interface EditorCardProps {
  /** The card image identifier to display. */
  imageIdentifier: string | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the previous image for visual smoothness. */
  previousImageIdentifier?: string | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the next image for visual smoothness. */
  nextImageIdentifier?: string | undefined;
  /** The string to display in the `Card` header. */
  cardHeaderTitle: string;
  /** An element (intended for use with a series of buttons) to include in the `Card` header.  */
  cardHeaderButtons?: ReactElement;
  /** An element (e.g. prev/next buttons) to display in the card footer. If not passed, no footer will be rendered. */
  cardFooter?: ReactElement;
  /** A callback function for when the `Card` (the HTML surrounding the image) is clicked. */
  cardOnClick?: React.MouseEventHandler<HTMLElement>;
  /** A callback function for when the card name is clicked. */
  nameOnClick?: React.MouseEventHandler<HTMLElement>;
  /** The `SearchQuery` specified when searching for this card. */
  searchQuery?: SearchQuery | undefined;
  /** Whether no search results were found when searching for `searchQuery` under the configured search settings. */
  noResultsFound: boolean;
  /** Whether to highlight this card by showing a glowing border around it. */
  highlight?: boolean;
}

/**
 * This component is a thin layer on top of `Card` that retrieves `CardDocument` items by their identifiers
 * from the Redux store (used in the project editor).
 * We have this layer because search results are returned as a list of image identifiers
 * (to minimise the quantity of data stored in Elasticsearch), so the full `CardDocument` items must be looked up.
 */
export function EditorCard({
  imageIdentifier,
  previousImageIdentifier,
  nextImageIdentifier,
  cardHeaderTitle,
  cardHeaderButtons,
  cardFooter,
  cardOnClick,
  nameOnClick,
  searchQuery,
  noResultsFound,
  highlight,
}: EditorCardProps) {
  //# region queries and hooks

  const maybeCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, imageIdentifier)
  );
  const maybePreviousCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, previousImageIdentifier)
  );
  const maybeNextCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, nextImageIdentifier)
  );

  //# endregion

  return (
    <Card
      maybeCardDocument={maybeCardDocument}
      maybePreviousCardDocument={maybePreviousCardDocument}
      maybeNextCardDocument={maybeNextCardDocument}
      cardHeaderTitle={cardHeaderTitle}
      cardHeaderButtons={cardHeaderButtons}
      cardFooter={cardFooter}
      cardOnClick={cardOnClick}
      nameOnClick={nameOnClick}
      searchQuery={searchQuery}
      noResultsFound={noResultsFound}
      highlight={highlight}
    />
  );
}

export const MemoizedEditorCard = memo(EditorCard);

interface DatedCardProps {
  cardDocument: CardDocument;
  headerDate: "created" | "modified";
}
/**
 * This component is a thin layer on top of `Card` for use in the What's New page.
 */
export function DatedCard({
  cardDocument,
  headerDate = "created",
}: DatedCardProps) {
  return (
    <Col>
      <MemoizedCard
        key={`new-cards-${cardDocument.identifier}`}
        maybeCardDocument={cardDocument}
        cardHeaderTitle={
          headerDate === "created"
            ? cardDocument.dateCreated
            : cardDocument.dateModified
        }
        noResultsFound={false}
      />
    </Col>
  );
}
