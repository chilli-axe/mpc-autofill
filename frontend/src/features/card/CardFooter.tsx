import React from "react";
import Button from "react-bootstrap/Button";

import { wrapIndex } from "@/common/utils";

export function CardFooter({
  searchResults,
  selectedImageIndex,
  selected,
  setSelectedImageFromIdentifier,
  handleShowGridSelector,
}: {
  searchResults: Array<string>;
  selectedImageIndex: number | undefined;
  selected: boolean;
  setSelectedImageFromIdentifier: { (image: string): void };
  handleShowGridSelector: { (): void };
}) {
  const setSelectedImageFromDelta = (delta: number): void => {
    // TODO: docstring
    if (selectedImageIndex !== undefined) {
      const image =
        searchResults[
          wrapIndex(selectedImageIndex + delta, searchResults.length)
        ];
      setSelectedImageFromIdentifier(image);
    }
  };

  return (
    <>
      {searchResults.length === 1 && (
        <p className="mpccard-counter text-center align-middle">
          1 / {searchResults.length}
        </p>
      )}
      {searchResults.length > 1 && (
        <>
          <Button
            variant={selected ? "info" : "outline-info"}
            className="mpccard-counter-btn"
            onClick={handleShowGridSelector}
          >
            {(selectedImageIndex ?? 0) + 1} / {searchResults.length}
          </Button>
          <div>
            <Button
              variant={selected ? "info" : "outline-info"}
              className="prev"
              onClick={() => setSelectedImageFromDelta(-1)}
            >
              &#10094;
            </Button>
            <Button
              variant={selected ? "info" : "outline-info"}
              className="next"
              onClick={() => setSelectedImageFromDelta(1)}
            >
              &#10095;
            </Button>
          </div>
        </>
      )}
    </>
  );
}
