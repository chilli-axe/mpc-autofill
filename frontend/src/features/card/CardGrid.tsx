/**
 * This component displays all `CardSlot`s in the project.
 */

import React from "react";
import Modal from "react-bootstrap/Modal";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { Back, Front } from "@/common/constants";
import { useAppSelector } from "@/common/types";
import { Spinner } from "@/components/Spinner";
import { MemoizedCardSlot } from "@/features/card/CardSlot";
import {
  selectIsProjectEmpty,
  selectProjectMembers,
} from "@/store/slices/projectSlice";
import { selectFrontsVisible } from "@/store/slices/viewSettingsSlice";

const CardGridDefaultBackground = styled.div`
  position: absolute;
  border: 1px darkgray solid;
  border-radius: 10px;
  height: 100%;
  width: 100%;
  box-shadow: 0 0 50px 0 black inset;
`;

const CardGridDefaultText = styled.div`
  position: relative;
  font-size: 1.25em;
  text-align: center;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-shadow: 0 4px 15px #000000;
  max-width: 90%;
`;

function CardGridDefault() {
  return (
    <CardGridDefaultBackground>
      <CardGridDefaultText>
        <p>Your project is empty at the moment.</p>
        <p>
          Use the <b>Add Cards</b> menu on the right to get started!
        </p>
      </CardGridDefaultText>
    </CardGridDefaultBackground>
  );
}

export function CardGrid() {
  //# region queries and hooks

  const fetchingCardData = useAppSelector(
    (state) => state.searchResults.status === "loading"
  );
  const isProjectEmpty = useAppSelector(selectIsProjectEmpty);
  const frontsVisible = useAppSelector(selectFrontsVisible);
  const projectMembers = useAppSelector(selectProjectMembers);

  //# endregion

  const cardSlotsFronts = [];
  const cardSlotsBacks = [];

  for (const [slot, slotProjectMember] of projectMembers.entries()) {
    cardSlotsFronts.push(
      <MemoizedCardSlot
        key={`${Front}-slot${slot}`}
        searchQuery={slotProjectMember.front?.query}
        face={Front}
        slot={slot}
      ></MemoizedCardSlot>
    );
    cardSlotsBacks.push(
      <MemoizedCardSlot
        key={`${Back}-slot${slot}`}
        searchQuery={slotProjectMember.back?.query}
        face={Back}
        slot={slot}
      ></MemoizedCardSlot>
    );
  }

  // TODO: should we aim to lift state up here and conditionally render rather than hide?
  return (
    <>
      {projectMembers.length == 0 && <CardGridDefault />}
      <Row
        xxl={4}
        lg={3}
        md={2}
        sm={1}
        xs={1}
        className="g-0"
        style={{ display: frontsVisible ? "" : "none" }}
      >
        {cardSlotsFronts}
      </Row>
      <Row
        xxl={4}
        lg={3}
        md={2}
        sm={1}
        xs={1}
        className="g-0"
        style={{ display: frontsVisible ? "none" : "" }}
      >
        {cardSlotsBacks}
      </Row>
      <Modal scrollable show={fetchingCardData && !isProjectEmpty} centered>
        <Modal.Header>
          <Modal.Title
            style={{ textAlign: "center", userSelect: "none" }}
            className="w-100"
          >
            Loading your cards...
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Spinner />
        </Modal.Body>
      </Modal>
    </>
  );
}
