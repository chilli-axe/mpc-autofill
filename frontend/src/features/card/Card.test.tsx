import {
  cardDocument1,
  cardDocument2,
  cardDocument3,
} from "@/common/test-constants";
import { renderWithProviders } from "@/common/test-utils";
import { MemoizedEditorCard } from "@/features/card/Card";

test("the html structure of a Card with a single document", () => {
  const rendered = renderWithProviders(
    <MemoizedEditorCard
      cardHeaderTitle="Slot 1"
      noResultsFound={false}
      imageIdentifier={cardDocument1.identifier}
    />,
    {
      preloadedState: {
        cardDocuments: {
          cardDocuments: {
            [cardDocument1.identifier]: cardDocument1,
          },
          status: "idle",
          error: null,
        },
      },
    }
  );
  expect(
    rendered.getByText("Slot 1").parentElement!.parentElement!
  ).toMatchSnapshot();
});

test("the html structure of a Card with two documents", () => {
  const rendered = renderWithProviders(
    <MemoizedEditorCard
      cardHeaderTitle="Slot 2"
      noResultsFound={false}
      imageIdentifier={cardDocument1.identifier}
      previousImageIdentifier={cardDocument2.identifier}
    />,
    {
      preloadedState: {
        cardDocuments: {
          cardDocuments: {
            [cardDocument1.identifier]: cardDocument1,
            [cardDocument2.identifier]: cardDocument2,
          },
          status: "idle",
          error: null,
        },
      },
    }
  );
  expect(
    rendered.getByText("Slot 2").parentElement!.parentElement!
  ).toMatchSnapshot();
});

test("the html structure of a Card with three documents", () => {
  const rendered = renderWithProviders(
    <MemoizedEditorCard
      cardHeaderTitle="Slot 3"
      noResultsFound={false}
      imageIdentifier={cardDocument1.identifier}
      previousImageIdentifier={cardDocument2.identifier}
      nextImageIdentifier={cardDocument3.identifier}
    />,
    {
      preloadedState: {
        cardDocuments: {
          cardDocuments: {
            [cardDocument1.identifier]: cardDocument1,
            [cardDocument2.identifier]: cardDocument2,
            [cardDocument3.identifier]: cardDocument3,
          },
          status: "idle",
          error: null,
        },
      },
    }
  );
  expect(
    rendered.getByText("Slot 3").parentElement!.parentElement!
  ).toMatchSnapshot();
});

test("the html structure of a Card with no search results", () => {
  const rendered = renderWithProviders(
    <MemoizedEditorCard
      imageIdentifier={undefined}
      cardHeaderTitle="Slot 1"
      noResultsFound={true}
      searchQuery={{ query: "My Invalid Query", cardType: "CARD" }}
    />,
    {
      preloadedState: {
        cardDocuments: {
          cardDocuments: {},
          status: "idle",
          error: null,
        },
      },
    }
  );
  expect(
    rendered.getByText("Slot 1").parentElement!.parentElement!
  ).toMatchSnapshot();
});
