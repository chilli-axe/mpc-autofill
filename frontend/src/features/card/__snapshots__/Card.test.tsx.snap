// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of a Card with a single document 1`] = `
.c1 {
  z-index: 1;
  opacity: 1;
}

.c1:hover {
  cursor: pointer;
}

.c0 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

<div
  class="mpccard mpccard-hover card"
>
  <div
    class="pb-0 text-center card-header"
  >
    <p
      class="mpccard-slot"
    >
      Slot 1
    </p>
  </div>
  <div>
    <div
      class="c0 rounded-lg shadow-lg ratio ratio-7x5"
    >
      <img
        alt="Card 1"
        class="c1 card-img card-img-fade-in"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="mb-0 text-center card-body"
    >
      <div
        class="mpccard-name card-subtitle h6"
      >
        Card 1
      </div>
      <div
        class="mpccard-spacing"
      >
        <p
          class="mpccard-source card-text"
        >
          Source 1 [1200 DPI]
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`the html structure of a Card with no search results 1`] = `
.c0 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

<div
  class="mpccard mpccard-hover card"
>
  <div
    class="pb-0 text-center card-header"
  >
    <p
      class="mpccard-slot"
    >
      Slot 1
    </p>
  </div>
  <div>
    <div
      class="c0 rounded-lg shadow-lg ratio ratio-7x5"
    >
      <img
        alt="Card not found"
        class="card-img card-img-fade-in"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=%2Fblank.png&w=3840&q=75"
        srcset="/_next/image?url=%2Fblank.png&w=640&q=75 640w, /_next/image?url=%2Fblank.png&w=750&q=75 750w, /_next/image?url=%2Fblank.png&w=828&q=75 828w, /_next/image?url=%2Fblank.png&w=1080&q=75 1080w, /_next/image?url=%2Fblank.png&w=1200&q=75 1200w, /_next/image?url=%2Fblank.png&w=1920&q=75 1920w, /_next/image?url=%2Fblank.png&w=2048&q=75 2048w, /_next/image?url=%2Fblank.png&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: 1;"
      />
    </div>
    <div
      class="mb-0 text-center card-body"
    >
      <div
        class="mpccard-name card-subtitle h6"
      >
        My Invalid Query
      </div>
      <div
        class="mpccard-spacing"
      >
        <p
          class="mpccard-source card-text"
        >
          Your search query
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`the html structure of a Card with three documents 1`] = `
.c2 {
  z-index: 0;
  opacity: 0;
}

.c1 {
  z-index: 1;
  opacity: 1;
}

.c1:hover {
  cursor: pointer;
}

.c0 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

<div
  class="mpccard mpccard-hover card"
>
  <div
    class="pb-0 text-center card-header"
  >
    <p
      class="mpccard-slot"
    >
      Slot 3
    </p>
  </div>
  <div>
    <div
      class="c0 rounded-lg shadow-lg ratio ratio-7x5"
    >
      <img
        alt="Card 1"
        class="c1 card-img card-img-fade-in"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
      <img
        alt="Card 2"
        class="c2 card-img"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
      <img
        alt="Card 3"
        class="c2 card-img"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="mb-0 text-center card-body"
    >
      <div
        class="mpccard-name card-subtitle h6"
      >
        Card 1
      </div>
      <div
        class="mpccard-spacing"
      >
        <p
          class="mpccard-source card-text"
        >
          Source 1 [1200 DPI]
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`the html structure of a Card with two documents 1`] = `
.c2 {
  z-index: 0;
  opacity: 0;
}

.c1 {
  z-index: 1;
  opacity: 1;
}

.c1:hover {
  cursor: pointer;
}

.c0 {
  z-index: 0;
  background: #4e5d6c;
  border: solid 0px black;
}

<div
  class="mpccard mpccard-hover card"
>
  <div
    class="pb-0 text-center card-header"
  >
    <p
      class="mpccard-slot"
    >
      Slot 2
    </p>
  </div>
  <div>
    <div
      class="c0 rounded-lg shadow-lg ratio ratio-7x5"
    >
      <img
        alt="Card 1"
        class="c1 card-img card-img-fade-in"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
      <img
        alt="Card 2"
        class="c2 card-img"
        data-nimg="fill"
        decoding="async"
        loading="lazy"
        src=""
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="mb-0 text-center card-body"
    >
      <div
        class="mpccard-name card-subtitle h6"
      >
        Card 1
      </div>
      <div
        class="mpccard-spacing"
      >
        <p
          class="mpccard-source card-text"
        >
          Source 1 [1200 DPI]
        </p>
      </div>
    </div>
  </div>
</div>
`;
